vowpalwabbit/plt.cc

```diff
@@ -0,0 +1,382 @@
+// Copyright (c) by respective owners including Yahoo!, Microsoft, and
+// individual contributors. All rights reserved. Released under a BSD (revised)
+// license as described in the file LICENSE.
+#include <cfloat>
+#include <cmath>
+#include <cstdio>
+#include <sstream>
+#include <vector>
+#include <unordered_map>
+#include <unordered_set>
+#include <algorithm>
+#include <queue>
+#include "reductions.h"
+#include "vw.h"
+
+using namespace VW::LEARNER;
+using namespace VW::config;
+
+namespace plt_ns
+{
+struct node
+{
+  uint32_t n;  // node number
+  float p;     // node probability
+
+  bool operator<(const node &r) const { return p < r.p; }
+};
+
+struct plt
+{
+  vw *all;
+
+  uint32_t k;     // number of labels
+  uint32_t t;     // number of tree nodes
+  uint32_t ti;    // number of internal nodes
+  uint32_t kary;  // kary tree
+
+  v_array<float> nodes_t;  // in case of sgd, this keeps stores individual t for each node
+
+  float threshold;
+  uint32_t top_k;
+  v_array<polyprediction> preds;  // for storing results of base.multipredict
+
+  // for measuring predictive performance
+  v_array<uint32_t> tp_at;  // true positives at (for precision and recall at)
+  uint32_t tp;
+  uint32_t fp;
+  uint32_t fn;
+  uint32_t t_count;   // number of all true labels (for recall at)
+  uint32_t ec_count;  // number of examples
+
+  plt()
+  {
+    nodes_t = v_init<float>();
+    preds = v_init<polyprediction>();
+    tp_at = v_init<uint32_t>();
+    tp = 0;
+    fp = 0;
+    fn = 0;
+    ec_count = 0;
+    t_count = 0;
+  }
+
+  ~plt()
+  {
+    nodes_t.delete_v();
+    preds.delete_v();
+    tp_at.delete_v();
+  }
+};
+
+inline void learn_node(plt &p, uint32_t n, single_learner &base, example &ec)
+{
+  if (!p.all->weights.adaptive)
+  {
+    p.all->sd->t = p.nodes_t[n];
+    p.nodes_t[n] += ec.weight;
+  }
+  base.learn(ec, n);
+}
+
+void learn(plt &p, single_learner &base, example &ec)
+{
+  MULTILABEL::labels multilabels = ec.l.multilabels;
+  MULTILABEL::labels preds = ec.pred.multilabels;
```

Comment:
For futureproofing these should like be moves?

Reply:
Like this?
```suggestion
  MULTILABEL::labels multilabels = std::move(ec.l.multilabels);
  MULTILABEL::labels preds = std::move(ec.pred.multilabels);
```

Reply:
Looks good!

Reply:
Added in all relevant places